### PR TITLE
Refactor notebook parsing

### PR DESCRIPTION
**Pull request by colombod**
We are adding a new package to expose support for parsing, reading and writing `InteractiveDocuments` to Jupyter Notebook or CodeSubmission (dib).

The Package`Microsoft.DotNet.Interactive.Documents` exposes all required api.

- Parsing from jupyter notebook : `var interactiveDocument = Notebook.Parse(content, KernelLanguageAliases)`
- Parsing from dib file : `var interactiveDocument =CodeSubmission.Parse( content,"csharp", KernelLanguageAliases);`
- Can read both formats from streams
- Can write both format to streams
- Can emit jupyter notebook content: `interactive.ToIpynbContent(newLine)`
- Can emit dib notebook content: `interactive.ToCodeSubmissionContent(newLine)`

### CODE REVIEW

File: src/Microsoft.DotNet.Interactive.Formatting/JsonElementExtensions.cs [hunk 1]
```diff
@@ -1,9 +1,9 @@
-﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System.Text.Json;
 
-namespace Microsoft.DotNet.Interactive.Extensions
+namespace Microsoft.DotNet.Interactive.Formatting
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Formatting/JsonElementExtensions.cs:6]: Does this actually need to be public? It seems unrelated to this package's purpose.

File: src/Microsoft.DotNet.Interactive.Notebooks.Tests/DibNotebookDocumentFileFormatTests.cs [hunk 2]
```diff
@@ -300,12 +295,13 @@ public void parsed_notebook_outputs_are_empty()
         public void notebook_can_be_parsed_from_different_extensions(string extension)
         {
             var fileName = $"notebook{extension}";
-            var notebook = ParseFromString(fileName, @"
+            var notebook = NotebookFileFormatHandler.Parse(fileName, @"
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Notebooks.Tests/DibNotebookDocumentFileFormatTests.cs:298]: These aren't strictly file formats, .dib in particular.

File: src/Microsoft.DotNet.Interactive.Notebooks/Notebook/NotebookFileFormatHandler.cs [hunk 3]
```diff
@@ -26,30 +25,35 @@ static NotebookFileFormatHandler()
...
+        public static NotebookDocument Parse(string fileName, string notebookContent, string defaultLanguage,
+            IDictionary<string, string> kernelLanguageAliases)
         {
             var extension = Path.GetExtension(fileName);
             switch (extension.ToLowerInvariant())
             {
                 case ".dib":
                 case ".dotnet-interactive":
-                    return ParseInteractiveNotebook(rawData, defaultLanguage, kernelLanguageAliases);
+                    return ParseInteractiveNotebook(notebookContent, defaultLanguage, kernelLanguageAliases);
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Notebooks/Notebook/NotebookFileFormatHandler.cs:36]: Any reason to keep support for the `.dotnet-interactive` file extension?

File: src/Microsoft.DotNet.Interactive.Notebooks/Notebook/NotebookFileFormatHandler.cs [hunk 4]
```diff
@@ -26,30 +25,35 @@ static NotebookFileFormatHandler()
...
-            var allLanguageAliases = ImmutableDictionary<string, string>.Empty
-                .AddRange(kernelLanguageAliases)
-                .Add("markdown", "markdown")
-                .Add("md", "markdown");
-
             var content = Encoding.UTF8.GetString(rawData);
+            return Parse(fileName, content, defaultLanguage, kernelLanguageAliases);
+        }
+
+        public static NotebookDocument ParseInteractiveNotebook(string content, string defaultLanguage, IDictionary<string, string> kernelLanguageAliases)
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Notebooks/Notebook/NotebookFileFormatHandler.cs:50]: There's some inconsistent naming here with `Parse` versus `ParseInteractiveNotebook`.

The model might be a little inside-out. Something more explicit and open-ended could be like:

```cs
NotebookDocument a = DibParser.Parse(...); // overloads supporting files or content
NotebookDocument a = JupyterNotebookParser.Parse(...); // overloads supporting files or content
NotebookDocument b = TryDotNetMarkdownParser.Parse(...); // overloads supporting files or content

```

The last one makes me wonder if `NotebookDocument` is the right name.

File: src/Microsoft.DotNet.Interactive.Formatting/JsonElementExtensions.cs (hunk 1 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Formatting/JsonElementExtensions.cs]: I moved that, that api is used in 2 places but no need to be on formatting

File: src/Microsoft.DotNet.Interactive.Notebooks.Tests/DibNotebookDocumentFileFormatTests.cs (hunk 2 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Notebooks.Tests/DibNotebookDocumentFileFormatTests.cs]: Don't sure I get this, .dib is a serialization format, no?

File: src/Microsoft.DotNet.Interactive.Notebooks/Notebook/NotebookFileFormatHandler.cs (hunk 3 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Notebooks/Notebook/NotebookFileFormatHandler.cs]: @brettfo ?

File: src/Microsoft.DotNet.Interactive.Tests/Server/SerializationTests.cs [hunk 5]
```diff
@@ -174,7 +174,7 @@ IEnumerable<KernelCommand> commands()
 
                 yield return new SendEditableCode("language", "code");
 
-                yield return new SerializeNotebook("interactive.ipynb", new InteractiveDocument(new[]
+                yield return new SerializeInteractiveDocument("interactive.ipynb", new InteractiveDocument(new[]
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Tests/Server/SerializationTests.cs:177]: "Parse" and "serialize" don't go really match. Maybe rename this to something like "WriteInteractiveDocument"?

Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Tests/Server/SerializationTests.cs:177]: Write / Read InteractiveDocument sounds good to me, @brettfo what is your take?

Comment (jonsequitur) [reply src/Microsoft.DotNet.Interactive.Tests/Server/SerializationTests.cs:177]: "Parse" makes sense to me but "serialize" doesn't.

File: src/Microsoft.DotNet.Interactive/Commands/SerializeInteractiveDocument.cs [hunk 6]
```diff
@@ -1,21 +1,19 @@
 ﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-using Microsoft.DotNet.Interactive.Notebook;
-
 namespace Microsoft.DotNet.Interactive.Commands
 {
-    public class SerializeNotebook : KernelCommand
+    public class SerializeInteractiveDocument : KernelCommand
```
Comment (colombod) [inline src/Microsoft.DotNet.Interactive/Commands/SerializeInteractiveDocument.cs:6]: @brettfo and @jonsequitur I have renamed these commands as they deal with our representation of an interactive document and they write it as dib or ipynb. Would like to pick your brain on these two commands. I think we have enough at this point in our object model to close this PR. Contract types are changed anyway (elements vs cells property for example) so I need to also align the typescript code in vscode extension (should be minimal work)

Comment (jonsequitur) [reply src/Microsoft.DotNet.Interactive/Commands/SerializeInteractiveDocument.cs:6]: Looks good to me.

File: src/Microsoft.DotNet.Interactive.Document.Tests/NotebookDocumentFileFormatTestsBase.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.DotNet.Interactive.Document.Tests
+{
+    public abstract class NotebookDocumentFileFormatTestsBase
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Document.Tests/NotebookDocumentFileFormatTestsBase.cs:9]: This naming pattern seems more verbose than necessary. In this context, "notebook" and "document" and "file" all mean more or less the same thing.

File: src/Microsoft.DotNet.Interactive.Document/NotebookFileFormatHandler.cs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.DotNet.Interactive.dib;
+using Microsoft.DotNet.Interactive.Ipynb;
+
+namespace Microsoft.DotNet.Interactive
+{
+    public static class NotebookFileFormatHandler
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Document/NotebookFileFormatHandler.cs:12]: I'd say the word "handler" is misleading here.

For context and discoverability, I'd maybe name this class `InteractiveDocument`.

File: src/Microsoft.DotNet.Interactive.Document/NotebookFileFormatHandler.cs [hunk 9]
```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.DotNet.Interactive.dib;
+using Microsoft.DotNet.Interactive.Ipynb;
+
+namespace Microsoft.DotNet.Interactive
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Document/NotebookFileFormatHandler.cs:10]: Namespace doesn't match the project.

File: src/Microsoft.DotNet.Interactive.Document/Microsoft.DotNet.Interactive.Document.csproj [hunk 10]
```diff
@@ -0,0 +1,25 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <LangVersion>preview</LangVersion>
+    <PackageId Condition="'$(PackageId)'==''">Microsoft.DotNet.Interactive.Document</PackageId>
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Document/Microsoft.DotNet.Interactive.Document.csproj:6]: "Documents" would be more conventional for a namespace.

File: src/Microsoft.DotNet.Interactive.Document/NotebookFileFormatHandler.cs [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+                case ".dotnet-interactive":
+                    return DibFile.Parse(content, defaultLanguage, kernelLanguageAliases);
+                case ".ipynb":
+                    return IpynbFile.Parse(content, kernelLanguageAliases);
+                default:
+                    throw new NotSupportedException($"Unable to parse a interactive document of type '{extension}'");
+            }
+        }
+
+        public static InteractiveDocument Read(string fileName, Stream stream, string defaultLanguage, IDictionary<string, string> kernelLanguageAliases)
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Document/NotebookFileFormatHandler.cs:30]: What's the different between `Read` and `Parse`?

File: src/Microsoft.DotNet.Interactive.Document/StringExtension.cs [hunk 12]
```diff
@@ -4,9 +4,9 @@
 using System.Collections.Generic;
 using System.Linq;
 
-namespace Microsoft.DotNet.Interactive.Notebook
+namespace Microsoft.DotNet.Interactive
 {
-    internal static class NotebookParsingExtensions
+    internal static class StringExtension
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Document/StringExtension.cs:9]:
```suggestion
    internal static class StringExtensions
```

File: src/Microsoft.DotNet.Interactive.ExtensionLab/RecordTranscriptExtension.cs [hunk 13]
```diff
@@ -49,7 +49,7 @@ public Task OnLoadAsync(Kernel kernel)
 
             KernelInvocationContext.Current?.Display(
                 new HtmlString(@"<details><summary>Use the <code>#!record</code> magic command to keep a transcript of the code you run.</summary>
-    <p>Once you enable transcripts using <code>#!record</code>, each code submission (including re-running cells) is recorded in the specified file. The JSON format used is the same format recognized by the .NET Interactive <code>stdio</code> and <code>http</code> APIs and can be used to replay a notebook session via automation.</p>
+    <p>Once you enable transcripts using <code>#!record</code>, each code submission (including re-running cells) is recorded in the specified file. The JSON format used is the same format recognized by the .NET Interactive <code>stdio</code> and <code>http</code> APIs and can be used to replay a interactive session via automation.</p>
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.ExtensionLab/RecordTranscriptExtension.cs:52]:
```suggestion
    <p>Once you enable transcripts using <code>#!record</code>, each code submission (including re-running cells) is recorded in the specified file. The JSON format used is the same format recognized by the .NET Interactive <code>stdio</code> and <code>http</code> APIs and can be used to replay an interactive session via automation.</p>
```

File: src/Microsoft.DotNet.Interactive.Jupyter.Tests/CompleteRequestHandlerTests.cs [hunk 14]
```diff
@@ -6,10 +6,8 @@
 
 using FluentAssertions;
 using FluentAssertions.Extensions;
-
+using Microsoft.DotNet.Interactive.Ipynb;
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Jupyter.Tests/CompleteRequestHandlerTests.cs:9]: I think `Jupyter` would make more sense here than `.ipynb`, no?

File: src/Microsoft.DotNet.Interactive.Tests/Server/SerializationTests.cs [hunk 15]
```diff
@@ -175,16 +174,16 @@ IEnumerable<KernelCommand> commands()
...
                 yield return new SendEditableCode("language", "code");
 
-                yield return new SerializeNotebook("notebook.ipynb", new NotebookDocument(new[]
+                yield return new SerializeInteractiveDocument("interactive.ipynb", new InteractiveDocument(new[]
                 {
-                    new NotebookCell("csharp", "user code", new NotebookCellOutput[]
+                    new InteractiveDocumentElement("csharp", "user code", new InteractiveDocumentOutputElement[]
                     {
-                        new NotebookCellDisplayOutput(new Dictionary<string, object>
+                        new InteractiveDocumentDisplayOutputElement(new Dictionary<string, object>
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Tests/Server/SerializationTests.cs:181]: Is "DisplayOutput" redundant?

File: src/Microsoft.DotNet.Interactive.Tests/Utility/CustomCommandTypes.cs [hunk 16]
```diff
@@ -8,7 +8,7 @@ namespace Microsoft.DotNet.Interactive.Tests.Utility
     public class CustomCommandTypes
     {
         // Multiple definitions of MyCommand simulate a user experimentally developing
-        // the command type in a notebook over time.
+        // the command type in a interactive over time.
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Tests/Utility/CustomCommandTypes.cs:11]: This doesn't make much sense.

File: src/Microsoft.DotNet.Interactive.Document.Tests/NotebookDocumentFileFormatTestsBase.cs (hunk 7 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Document.Tests/NotebookDocumentFileFormatTestsBase.cs:9]: makes sense

File: src/Microsoft.DotNet.Interactive.Documents/InteractiveDocument.cs [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace Microsoft.DotNet.Interactive.Documents
+{
+    public partial class InteractiveDocument
+    {
+        public InteractiveDocumentElement[] Elements { get; }
+
+        public InteractiveDocument(InteractiveDocumentElement[] elements)
+        {
+            Elements = elements;
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Documents/InteractiveDocument.cs:12]: Needs a null check.

File: src/Microsoft.DotNet.Interactive.Documents/Jupyter/InputCellMetadata.cs [hunk 18]
```diff
@@ -1,11 +1,11 @@
 ﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-
 using System.Text.Json.Serialization;
 
-namespace Microsoft.DotNet.Interactive.Notebook
+namespace Microsoft.DotNet.Interactive.Ipynb
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Documents/Jupyter/InputCellMetadata.cs:6]:
```suggestion
namespace Microsoft.DotNet.Interactive.Documents.Jupyter
```

File: src/Microsoft.DotNet.Interactive.Documents/Jupyter/Notebook.cs [hunk 19]
```diff
@@ -0,0 +1,278 @@
...
+
+        public static async Task<InteractiveDocument> ReadAsync(Stream stream,
+            IDictionary<string, string> kernelLanguageAliases)
+        {
+            using var reader = new StreamReader(stream, Encoding);
+            var content = await reader.ReadToEndAsync();
+            return Parse(content, kernelLanguageAliases);
+        }
+
+        private static List<string> GetTextLines(JsonElement? jsonElement)
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Documents/Jupyter/Notebook.cs:138]: Why allow null? I would think that would typically only result from a programming error.

File: src/Microsoft.DotNet.Interactive.Documents/Jupyter/Notebook.cs [hunk 20]
```diff
@@ -0,0 +1,278 @@
...
+        }
+
+        public static void Write(Documents.InteractiveDocument interactive, string newline, Stream stream)
+        {
+            using var writer = new StreamWriter(stream, Encoding, 1024, true);
+            Write(interactive, newline, writer);
+            writer.Flush();
+        }
+
+        public static string ToIpynbContent(this Documents.InteractiveDocument interactive, string newline = "\n")
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Documents/Jupyter/Notebook.cs:163]:
```suggestion
        public static string ToJupyterNotebookContent(this Documents.InteractiveDocument interactive, string newline = "\n")
```

...maybe?

File: src/Microsoft.DotNet.Interactive.Documents/InteractiveDocument.cs (hunk 17 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Documents/InteractiveDocument.cs:12]: done

File: src/Microsoft.DotNet.Interactive.Documents/Jupyter/Notebook.cs (hunk 19 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Documents/Jupyter/Notebook.cs:138]: the element is fetched by trying to get it from the document, it can be null when not found

Comment (jonsequitur) [reply src/Microsoft.DotNet.Interactive.Documents/Jupyter/Notebook.cs:138]: When is it not found? Does this indicate a badly formed document? Should we be emitting a cell at all?

File: src/Microsoft.DotNet.Interactive.Jupyter.Tests/CompleteRequestHandlerTests.cs (hunk 14 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Jupyter.Tests/CompleteRequestHandlerTests.cs]: Done

File: src/Microsoft.DotNet.Interactive.Tests/Server/SerializationTests.cs (hunk 15 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Tests/Server/SerializationTests.cs]: renamed types

File: src/Microsoft.DotNet.Interactive.Documents/Jupyter/Notebook.cs (hunk 19 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Documents/Jupyter/Notebook.cs:138]: We should emit the empty lines as we found a cell but the source element wasn't present.

Comment (brettfo) [reply src/Microsoft.DotNet.Interactive.Documents/Jupyter/Notebook.cs:138]: This is us being overly cautious in the scenario that the JSON is reporting that we're in a new cell, but there's no `"source"` property.  Rather than add a ton of `null` checks, it's easier to make these accept a nullable and default to empty strings/collections.

File: src/Microsoft.DotNet.Interactive.Document/Microsoft.DotNet.Interactive.Document.csproj (hunk 10 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Document/Microsoft.DotNet.Interactive.Document.csproj]: done

File: src/Microsoft.DotNet.Interactive.Documents/Jupyter/InputCellMetadata.cs (hunk 18 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Documents/Jupyter/InputCellMetadata.cs]: done

File: src/Microsoft.DotNet.Interactive.Notebooks/Notebook/NotebookFileFormatHandler.cs (hunk 3 above)
Comment (jonsequitur) [reply src/Microsoft.DotNet.Interactive.Notebooks/Notebook/NotebookFileFormatHandler.cs]: Could you open an issue for this and we'll revisit?

**Review by jonsequitur — APPROVED** [review]
